### DISCUSSION TITLE
Expand string to bool conversion to include more strings

### DISCUSSION THREAD

**Discussion by mcastorina** [vlang/v#11822] [Ideas]
Go's string to bool conversion uses these values.

```
"1", "t", "T", "true", "TRUE", "True"
"0", "f", "F", "false", "FALSE", "False"
```

I think V should use the same.

**Comment by ralyodio** [comment]
agreed

**Comment by zakuro9715** [comment]
I don't completely agree. There is no de facto standard for string to bool conversion. Go is not always right and we don't have to same as Go.

But current behavior is not intuitive for me. We should reconsider for this.

```
>>> ''.bool()
false
>>> '1'.bool()
false
>>> 'a'.bool()
false
>>> 'true'.bool()
true
>>> 'True'.bool()
false
>>>
```

**Comment by ralyodio** [comment]
If we are going to allow `"string".bool()` i think anything except "false" "1" etc should be allowed
so i can do like:
```
name := "chovy"
nameExists := name.bool()

if nameExists {
   // name is defined 
}
``

**Comment by mcastorina** [comment]
> Go is not always right and we don't have to same as Go.

Agree here, but in this case I think they have good values.

> i think anything except "false" "1" etc should be allowed

I disagree with this because it's even less intuitive. Some values will be "defined" but if it's "false" then it's "not defined"?

```v
name := "false"
if name.bool() {
     // code does not run here
}
```

It's better to just use `if name.len > 0` for this.

**Comment by dumblob** [comment]
I think this should behave **exactly the same** as `if 'any_string_here' { ... } else { ... }` (and if this implicit conversion is not wanted, then I'd say we might get inspiration from C and maybe sh/Python/Ruby/Lua/...).

**Comment by JalonSolov** [comment]
V is much more opinionated about bools, so making this "conversion" opinionated as well makes sense.  I would only allow

"t", "true" (in _any_ case)
"f", "false" (in _any_ case)

Allowing '1' or '0' would imply you can assign `1` or `0` to a bool variable, which V doesn't allow.

The question is:  What to do if it isn't one of those above?  Warning/error?

**Comment by mcastorina** [comment]
It could return an optional, but that's inconsistent with `str.int()`, so I would say have two functions: `str.bool()` returns `true` if the string is in the "true" list, false otherwise and `parse_bool(str)` returns a `?bool`.

**Comment by dumblob** [comment]
Thinking about this there was already a discussion about `bool` versus `int` and the decision was to **always** use the "ternary expression" (`return if myint == 0 { true } else { false }`) instead of `bool( myint )`.

Now the **big question** is, why would we want to introduce inconsistency (and a really hairy one) into *V* and allow e.g. `str` to be directly castable to `bool`.

It feels like a rabbit hole when trying to come up with an easy solution:

Strings **are** considered UTF-8 in *V* and from that follows we'd need to support all (real and artificial and slangs etc.) languages in the world (if you say it's nonsense then believe me it's not - because if languages are not supported, then you can't just write `bool( input_string )` because it wouldn't behave as you'd expect (not speaking about blank characters, many types of spaces, etc.).

So you would anyway use the "ternary expression" in which case supporting direct casting of `string` to `bool` is pure useless bloat in otherwise sleek & lean *V*.

**Comment by mcastorina** [comment]
I strongly disagree with allowing non-bool types in `if` statements. It goes against V's safety by adding ambiguity to the language. It reads too much like "if x is defined", but in V all variables are always defined (initialized).

```v
x := []string{}
if !x { }         // not readable
if x.len == 0 { } // readable
```

**Comment by dumblob** [comment]
I apologize - it was just a demonstration and I did not mean to introduce `if any_non_bool_variable {` into the language. I'll reword the comment.

**Comment by mcastorina** [comment]
@dumblob ah I misunderstood because I think we are discussing two separate ideas: `bool(str)` and `str.bool()`

I agree with you that we should not allow `bool(str)`. `str.bool()` on the other hand, is parsing `str` into a boolean value (similar to how `str.int()` parses a string into an integer value.

Currently `str.bool()` returns true for a few values like `'true'` and `'t'`, and false otherwise. I'm proposing we extend that list, and possibly add a `strconv.parse_bool` function that returns a `?bool`.

**Comment by dumblob** [comment]
I'm sorry if I missed something important in *V* development - I'm unfortunately not following it's whole development due to lack of time. My understanding was, that if any type provides `str()` or `bool()` or `int()` or `int64()` etc. method, then there is syntactic sugar which automatically allows to use it as casting syntax in which case my comment would still hold.

In other cases, just ignore that comment of mine :wink:.

**Comment by ntrel** [comment]
> there was already a discussion about bool versus int and the decision was to always use the "ternary expression"

@dumblob That would break V's support for branchless programming where you need to cast bool to int to avoid a costly branch instruction.

**Comment by JalonSolov** [comment]
bools are not ints (regardless of the current implementation).  If you want to 'cast' a bool to an int, do something like
```v
mybool := if myint == 0 {false} else {true}
```

**Comment by ntrel** [comment]
@JalonSolov did you read my comment? You can't use `if` for branchless programming, which is important for manual optimization because no optimizers are as good as a human at avoiding jump instructions.

**Comment by JalonSolov** [comment]
I understand you can't use if in branchless programming.  That does not change the fact that ints and bools are separate things, and can't be cast to each other.

**Comment by ntrel** [comment]
@JalonSolov Says who? We deliberately support casting from bool to int, to support manual branchless optimizations.

**Comment by dumblob** [comment]
>@dumblob That would break V's support for branchless programming where you need to cast int to bool to avoid a costly branch instruction.

Of course - but it was one of the few **main** arguments why to introduce `boot` at all. @medvednikov any idea how to solve this?

I for myself consider "branchless programming" largely unaffected by this - I will yet need to see branchless code operating upon `bool`s.

**Comment by dumblob** [comment]
I think we have more trouble than I initially thought - `bool` is [allowed to be casted](https://github.com/vlang/v/pull/9676 ) to ints :open_mouth:.
